src/PythonMigrationViewExtension/PythonMigrationViewExtension.cs

```diff
@@ -125,23 +129,85 @@ private void OnNodeAdded(Graph.Nodes.NodeModel obj)
         {
             if (Configuration.DebugModes.IsEnabled("Python2ObsoleteMode")
                 && !NotificationTracker.ContainsKey(CurrentWorkspace.Guid)
-                && GraphPythonDependencies.IsIronPythonNode(obj))
+                && PythonDependencies.IsIronPythonNode(obj))
             {
                 LogIronPythonNotification();
             }
+
+            if (obj is PythonNodeBase)
+            {
+                SubscribeToPythonNodeEvents(obj as PythonNodeBase);
+            }
+        }
+
+        private void OnNodeRemoved(Graph.Nodes.NodeModel obj)
+        {
+            if (!PythonDependencies.IsIronPythonNode(obj) &&
+                !PythonDependencies.IsCPythonNode(obj))
+                return;
+
+            PythonDependencies.RemovePythonNode(obj);
+            if (!(obj is PythonNodeBase pythonNode))
+                return;
+
+            UnSubscribePythonNodeEvents(pythonNode);
+
         }
 
         private void OnCurrentWorkspaceChanged(IWorkspaceModel workspace)
         {
+            SubscribeToWorkspaceEvents();
             NotificationTracker.Remove(CurrentWorkspace.Guid);
             CurrentWorkspace = workspace as WorkspaceModel;
+            PythonDependencies = new GraphPythonDependencies(LoadedParams);
             if (Configuration.DebugModes.IsEnabled("Python2ObsoleteMode")
                 && !Models.DynamoModel.IsTestMode
                 && PythonDependencies.ContainsIronPythonDependencies())
             {
                 LogIronPythonNotification();
                 DisplayIronPythonDialog();
             }
+
+            if (PythonDependencies.GraphPythonNodes.Any())
+            {
+                PythonDependencies.GraphPythonNodes.ForEach(x => SubscribeToPythonNodeEvents(x));
+            }
+        }
+
+        private void SubscribeToDynamoEvents()
+        {
+            LoadedParams.CurrentWorkspaceChanged += OnCurrentWorkspaceChanged;
+            DynamoViewModel.Model.Logger.NotificationLogged += OnNotificationLogged;
+            SubscribeToWorkspaceEvents();
+        }
+
+        private void SubscribeToWorkspaceEvents()
+        {
+            DynamoViewModel.CurrentSpaceViewModel.Model.NodeAdded += OnNodeAdded;
+            DynamoViewModel.CurrentSpaceViewModel.Model.NodeRemoved += OnNodeRemoved;
+        }
+
+        private void SubscribeToPythonNodeEvents(PythonNodeBase node)
+        {
+            node.MigrationAssistantRequested += OnMigrationAssistantRequested;
+        }
+
+        private void UnSubscribePythonNodeEvents(PythonNodeBase node)
+        {
+            node.MigrationAssistantRequested -= OnMigrationAssistantRequested;
+        }
+
+        private void UnSubscribeWorkspaceEvents()
+        {
+            DynamoViewModel.CurrentSpaceViewModel.Model.NodeAdded -= OnNodeAdded;
+            DynamoViewModel.CurrentSpaceViewModel.Model.NodeAdded -= OnNodeRemoved;
+        }
+
+        private void UnsubscribeEvents()
+        {
```

Comment:
we should go through all Python nodes in workspace and unsubscribe from migration event here as well, just in case our list of nodes is not in sync with the workspace